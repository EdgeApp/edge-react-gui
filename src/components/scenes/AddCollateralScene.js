// @flow

import s from '../../locales/strings.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { ManageCollateralScene } from './ManageCollateralScene.js'

type Props = {
  navigation: NavigationProp<'addCollateralScene'>,
  route: RouteProp<'addCollateralScene'>
}

export const AddCollateralScene = (props: Props) => {
  const { navigation, route } = props
  const { borrowEngine } = route.params

  return ManageCollateralScene({
    borrowEngine,
    defaultTokenId: borrowEngine.collaterals[0].tokenId,
    action: async req => await borrowEngine.deposit(req),
    actionWallet: 'fromWallet',
    ltvType: 'collaterals',
    ltvChange: 'decrease',

    showExchangeRateTile: true,
    showTotalDebtTile: true,
    showTotalCollateralTile: true,

    headerText: s.strings.loan_add_collateral,
    goBack: () => navigation.pop()
  })
}
