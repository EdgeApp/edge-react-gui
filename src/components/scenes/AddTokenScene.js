// @flow

import * as React from 'react'
import { ActivityIndicator, Alert, ScrollView, StyleSheet, View } from 'react-native'

import { addNewToken } from '../../actions/AddTokenActions.js'
import { MAX_TOKEN_CODE_CHARACTERS } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import { FormattedText as Text } from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { connect } from '../../types/reactRedux.js'
import { type RouteProp } from '../../types/routerTypes.js'
import type { CustomTokenInfo, GuiWallet } from '../../types/types.js'
import { scale } from '../../util/scaling.js'
import { decimalPlacesToDenomination } from '../../util/utils.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

type OwnProps = {
  route: RouteProp<'addToken'>
}

type DispatchProps = {
  addNewToken: (walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, type: string) => void
}

type StateProps = {
  addTokenPending: boolean,
  currentCustomTokens: CustomTokenInfo[],
  wallet: GuiWallet
}

type State = {
  currencyName: string,
  currencyCode: string,
  contractAddress: string,
  decimalPlaces: string,
  enabled?: boolean
}

type Props = OwnProps & StateProps & DispatchProps

class AddToken extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { route } = this.props
    const { currencyName = '', currencyCode = '', contractAddress = '', decimalPlaces = '' } = route.params
    this.state = {
      currencyName,
      currencyCode,
      contractAddress,
      decimalPlaces
    }
  }

  render() {
    const { addTokenPending } = this.props
    return (
      <SceneWrapper background="body">
        <ScrollView style={styles.container}>
          <View style={styles.instructionalArea}>
            <Text style={styles.instructionalText}>{s.strings.addtoken_top_instructions}</Text>
          </View>
          <View style={styles.nameArea}>
            <FormField
              value={this.state.currencyName}
              onChangeText={this.onChangeName}
              autoCapitalize="words"
              autoFocus
              label={s.strings.addtoken_name_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.currencyCodeArea}>
            <FormField
              value={this.state.currencyCode}
              onChangeText={this.onChangeCurrencyCode}
              autoCapitalize="characters"
              label={s.strings.addtoken_currency_code_input_text}
              returnKeyType="done"
              autoCorrect={false}
              maxLength={MAX_TOKEN_CODE_CHARACTERS}
            />
          </View>
          <View style={styles.contractAddressArea}>
            <FormField
              value={this.state.contractAddress}
              onChangeText={this.onChangeContractAddress}
              label={s.strings.addtoken_contract_address_input_text}
              returnKeyType="done"
              autoCorrect={false}
            />
          </View>
          <View style={styles.decimalPlacesArea}>
            <FormField
              value={this.state.decimalPlaces}
              onChangeText={this.onChangeDecimalPlaces}
              label={s.strings.addtoken_denomination_input_text}
              returnKeyType="done"
              autoCorrect={false}
              keyboardType="numeric"
            />
          </View>
          <View style={styles.buttonsArea}>
            <PrimaryButton style={styles.saveButton} onPress={this._onSave}>
              {addTokenPending ? <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} /> : <PrimaryButton.Text>{s.strings.string_save}</PrimaryButton.Text>}
            </PrimaryButton>
          </View>
          <View style={styles.bottomPaddingForKeyboard} />
        </ScrollView>
      </SceneWrapper>
    )
  }

  onChangeName = (input: string) => {
    this.setState({
      currencyName: input
    })
  }

  onChangeCurrencyCode = (input: string) => {
    this.setState({
      currencyCode: input
    })
  }

  onChangeDecimalPlaces = (input: string) => {
    this.setState({
      decimalPlaces: input
    })
  }

  onChangeContractAddress = (input: string) => {
    this.setState({
      contractAddress: input.trim()
    })
  }

  _onSave = () => {
    const currencyCode = this.state.currencyCode.toUpperCase()
    this.setState(
      {
        currencyCode
      },
      () => {
        const { route, currentCustomTokens, wallet, addNewToken } = this.props
        const { currencyName, decimalPlaces, contractAddress } = this.state
        const { walletId } = route.params

        const currentCustomTokenIndex = currentCustomTokens.findIndex(item => item.currencyCode === currencyCode)
        const metaTokensIndex = wallet.metaTokens.findIndex(item => item.currencyCode === currencyCode)
        // if token is hard-coded into wallets of this type
        if (metaTokensIndex >= 0) Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        // if that token already exists and is visible (ie not deleted)
        if (currentCustomTokenIndex >= 0 && currentCustomTokens[currentCustomTokenIndex].isVisible !== false) {
          Alert.alert(s.strings.manage_tokens_duplicate_currency_code)
        } else {
          if (currencyName && currencyCode && decimalPlaces && contractAddress) {
            const denomination = decimalPlacesToDenomination(decimalPlaces)
            addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, wallet.type)
          } else {
            Alert.alert(s.strings.addtoken_invalid_information)
          }
        }
      }
    )
  }
}

const rawStyles = {
  container: {
    flex: 1,
    paddingHorizontal: scale(20)
  },

  instructionalArea: {
    paddingVertical: scale(16),
    paddingHorizontal: scale(20)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center'
  },

  nameArea: {
    height: scale(70)
  },
  currencyCodeArea: {
    height: scale(70)
  },
  contractAddressArea: {
    height: scale(70)
  },
  decimalPlacesArea: {
    height: scale(70)
  },
  buttonsArea: {
    marginTop: scale(16),
    height: scale(52),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    alignSelf: 'flex-end',
    paddingVertical: scale(4)
  },
  saveButton: {
    flex: 1,
    marginLeft: scale(2),
    backgroundColor: THEME.COLORS.SECONDARY,
    borderRadius: 3
  },
  bottomPaddingForKeyboard: {
    height: scale(300)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const AddTokenScene = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    addTokenPending: state.ui.wallets.addTokenPending,
    currentCustomTokens: state.ui.settings.customTokens,
    wallet: state.ui.wallets.byId[params.walletId]
  }),
  dispatch => ({
    addNewToken(walletId: string, currencyName: string, currencyCode: string, contractAddress: string, denomination: string, walletType: string) {
      dispatch(addNewToken(walletId, currencyName, currencyCode, contractAddress, denomination, walletType))
    }
  })
)(AddToken)
