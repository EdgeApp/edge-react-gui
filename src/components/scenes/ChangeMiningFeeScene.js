// @flow

import { type EdgeSpendTarget } from 'edge-core-js/types'
import * as React from 'react'
import { ScrollView, StyleSheet, Text, TouchableWithoutFeedback, View } from 'react-native'
import EntypoIcon from 'react-native-vector-icons/Entypo'

import { sendConfirmationUpdateTx } from '../../actions/SendConfirmationActions.js'
import { FEE_STRINGS } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import { getGuiMakeSpendInfo } from '../../modules/UI/scenes/SendConfirmation/selectors.js'
import { dayText, nightText } from '../../styles/common/textStyles.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { connect } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { type FeeOption } from '../../types/types.js'
import { FormField } from '../common/FormField.js'
import { SceneWrapper } from '../common/SceneWrapper.js'
import { showError } from '../services/AirshipInstance.js'

type OwnProps = {
  navigation: NavigationProp<'changeMiningFee'>,
  route: RouteProp<'changeMiningFee'>
}

type StateProps = {
  networkFeeOption?: FeeOption,
  customNetworkFee?: Object,
  spendTargets?: EdgeSpendTarget[],
  maxSpendSet: boolean
}

type DispatchProps = {
  onSubmit: (networkFeeOption: string, customNetworkFee: Object, walletId: string, currencyCode?: string) => void
}

type Props = OwnProps & StateProps & DispatchProps

type State = {
  networkFeeOption: FeeOption,
  customNetworkFee: Object
}

export class ChangeMiningFee extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    const { networkFeeOption = 'standard', customNetworkFee = {} } = props
    const customFormat = this.getCustomFormat()

    if (customFormat != null && Object.keys(customNetworkFee).length !== customFormat.length) {
      // Reset the custom fees if they don't match the format:
      const defaultCustomFee = {}
      for (const key of customFormat) defaultCustomFee[key] = ''
      this.state = { networkFeeOption, customNetworkFee: defaultCustomFee }
    } else {
      // Otherwise, use the custom fees from before:
      this.state = { networkFeeOption, customNetworkFee }
    }
  }

  getCustomFormat(): string[] | void {
    const { route } = this.props
    const { wallet } = route.params
    if (wallet.currencyInfo.defaultSettings != null) {
      const { customFeeSettings } = wallet.currencyInfo.defaultSettings
      return customFeeSettings
    }
  }

  onSubmit = () => {
    const { networkFeeOption, customNetworkFee } = this.state
    const { spendTargets = [], maxSpendSet, route } = this.props
    const { currencyCode, wallet } = route.params
    const testSpendInfo = {
      spendTargets: spendTargets.map(spendTarget => ({
        ...spendTarget,
        nativeAmount: maxSpendSet || spendTarget.nativeAmount === '' ? '0' : spendTarget.nativeAmount
      })),
      networkFeeOption,
      customNetworkFee,
      currencyCode
    }
    wallet
      .makeSpend(testSpendInfo)
      .then(() => {
        this.props.onSubmit(networkFeeOption, customNetworkFee, wallet.id, currencyCode)
        this.props.navigation.pop()
      })
      .catch(e => {
        let message = e.message
        if (e.name === 'ErrorBelowMinimumFee') message = `${s.strings.invalid_custom_fee} ${e.message}`
        showError(message)
      })
    // this.props.navigation.push('SomewhereElse')
  }

  render() {
    const customFormat = this.getCustomFormat()

    return (
      <SceneWrapper background="body" hasTabs={false} avoidKeyboard>
        <ScrollView style={styles.content}>
          {this.renderRadioRow('high', s.strings.mining_fee_high_label_choice)}
          {this.renderRadioRow('standard', s.strings.mining_fee_standard_label_choice)}
          {this.renderRadioRow('low', s.strings.mining_fee_low_label_choice)}
          {customFormat != null ? this.renderRadioRow('custom', s.strings.mining_fee_custom_label_choice) : null}
          {customFormat != null ? this.renderCustomFee(customFormat) : null}
          {this.renderFeeWarning()}
          <PrimaryButton onPress={this.onSubmit} style={styles.saveButton}>
            <PrimaryButton.Text>{s.strings.save}</PrimaryButton.Text>
          </PrimaryButton>
        </ScrollView>
      </SceneWrapper>
    )
  }

  renderRadioRow(value: FeeOption, label: string) {
    const { networkFeeOption } = this.state

    return (
      <TouchableWithoutFeedback onPress={() => this.setState({ networkFeeOption: value })}>
        <View style={styles.radioRow}>
          <View style={[styles.radio, networkFeeOption === value ? styles.selected : null]} />
          <Text style={dayText('row-left')}>{label}</Text>
        </View>
      </TouchableWithoutFeedback>
    )
  }

  renderCustomFee(customFormat: string[]): React.Node {
    const { networkFeeOption, customNetworkFee } = this.state
    if (networkFeeOption !== 'custom') return null

    return (
      <View style={styles.customArea}>
        {customFormat.map(key => (
          <FormField
            key={key}
            keyboardType="numeric"
            onChangeText={text =>
              this.setState({
                customNetworkFee: { ...customNetworkFee, [key]: text }
              })
            }
            value={customNetworkFee[key]}
            label={FEE_STRINGS[key] || key}
          />
        ))}
      </View>
    )
  }

  renderFeeWarning() {
    const { networkFeeOption } = this.state
    if (networkFeeOption !== 'custom' && networkFeeOption !== 'low') return null

    return (
      <View style={styles.warningBox}>
        <EntypoIcon name="warning" color={THEME.COLORS.WHITE} size={THEME.rem(1.4)} />
        <Text style={nightText('small')}>{s.strings.warning_low_or_custom_fee}</Text>
      </View>
    )
  }
}

const rawStyles = {
  content: {
    flexGrow: 1,
    backgroundColor: THEME.COLORS.WHITE,
    padding: THEME.rem(1.4)
  },

  // Radio input:
  radioRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: THEME.rem(1)
  },
  radio: {
    borderRadius: THEME.rem(0.5),
    marginRight: THEME.rem(0.5),
    width: THEME.rem(1),
    height: THEME.rem(1),
    borderWidth: THEME.rem(1 / 16),
    borderColor: THEME.COLORS.GRAY_2
  },
  selected: {
    borderColor: THEME.COLORS.ACCENT_BLUE,
    backgroundColor: THEME.COLORS.ACCENT_BLUE
  },

  // Custom fee area:
  customArea: {
    marginBottom: THEME.rem(1)
  },

  // Warning box:
  warningBox: {
    padding: THEME.rem(0.5),

    backgroundColor: THEME.COLORS.ACCENT_ORANGE,
    borderRadius: THEME.rem(0.5),

    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'space-between'
  },

  saveButton: {
    marginTop: THEME.rem(1.25)
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const ChangeMiningFeeScene = connect<StateProps, DispatchProps, OwnProps>(
  state => ({
    networkFeeOption: getGuiMakeSpendInfo(state).networkFeeOption,
    customNetworkFee: getGuiMakeSpendInfo(state).customNetworkFee,
    spendTargets: getGuiMakeSpendInfo(state).spendTargets,
    maxSpendSet: state.ui.scenes.sendConfirmation.maxSpendSet
  }),
  dispatch => ({
    onSubmit(networkFeeOption: string, customNetworkFee: Object, walletId: string, currencyCode?: string) {
      dispatch(sendConfirmationUpdateTx({ networkFeeOption, customNetworkFee }, true, walletId, currencyCode, true))
    }
  })
)(ChangeMiningFee)
