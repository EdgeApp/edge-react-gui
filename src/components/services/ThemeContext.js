// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
