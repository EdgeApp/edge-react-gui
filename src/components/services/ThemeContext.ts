import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
