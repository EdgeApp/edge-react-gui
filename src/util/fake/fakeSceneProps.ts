import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
