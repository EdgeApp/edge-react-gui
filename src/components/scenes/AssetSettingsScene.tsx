import * as React from 'react'
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome'

import { CURRENCY_SETTINGS_KEYS } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useSelector } from '../../types/reactRedux'
import { EdgeAppSceneProps } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { CryptoIcon } from '../icons/CryptoIcon'
import { showToast } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { SettingsTappableRow } from '../settings/SettingsTappableRow'

interface Props extends EdgeAppSceneProps<'assetSettings'> {}

export function AssetSettingsScene(props: Props) {
  const { navigation } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const { currencyWallets } = account

  /**
   * Iterates through all the user's currency wallets, updates the enabled token
   * IDs to include any newly detected tokens, and then displays a toast message
   * indicating whether any detected tokens were enabled.
   */
  const handleRescanTokens = useHandler(async () => {
    let isTokensDetected = false
    for (const wallet of Object.values(currencyWallets)) {
      const detectedDisabledTokens = wallet.detectedTokenIds.filter(dt => !wallet.enabledTokenIds.includes(dt))

      if (detectedDisabledTokens.length > 0) {
        await wallet.changeEnabledTokenIds([...wallet.enabledTokenIds, ...detectedDisabledTokens])
        isTokensDetected = true
      }
    }

    if (isTokensDetected) {
      showToast(lstrings.settings_enable_detected_tokens_toast)
    } else {
      showToast(lstrings.settings_no_detected_tokens_toast)
    }
  })

  return (
    <SceneWrapper scroll>
      <SettingsTappableRow key="detectTokens" label={lstrings.settings_detect_tokens} onPress={handleRescanTokens}>
        <FontAwesomeIcon style={styles.icon} name="refresh" size={theme.rem(1.25)} color={theme.iconTappable} />
      </SettingsTappableRow>
      {CURRENCY_SETTINGS_KEYS.map(pluginId => {
        if (account.currencyConfig[pluginId] == null) return null
        const { currencyInfo } = account.currencyConfig[pluginId]
        const { displayName } = currencyInfo
        const onPress = () =>
          navigation.navigate('currencySettings', {
            currencyInfo
          })

        return (
          <SettingsTappableRow key={pluginId} label={displayName} onPress={onPress}>
            <CryptoIcon marginRem={[0.5, 0, 0.5, 0.5]} pluginId={pluginId} tokenId={null} sizeRem={1.25} />
          </SettingsTappableRow>
        )
      })}
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  icon: {
    marginLeft: theme.rem(0.6)
  }
}))
