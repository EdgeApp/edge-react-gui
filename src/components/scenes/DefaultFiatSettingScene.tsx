import { FlashList } from '@shopify/flash-list'
import * as React from 'react'
import { Alert, Keyboard, View } from 'react-native'
import FastImage from 'react-native-fast-image'
import { cacheStyles } from 'react-native-patina'

import { setDefaultFiatRequest } from '../../actions/SettingsActions'
import { FIAT_COUNTRY } from '../../constants/CountryConstants'
import s from '../../locales/strings'
import { getDefaultFiat } from '../../selectors/SettingsSelectors'
import { connect } from '../../types/reactRedux'
import { NavigationProp } from '../../types/routerTypes'
import { Theme } from '../../types/Theme'
import { FlatListItem, GuiFiatType } from '../../types/types'
import { scale } from '../../util/scaling'
import { getSupportedFiats } from '../../util/utils'
import { SceneWrapper } from '../common/SceneWrapper'
import { ThemeProps, withTheme } from '../services/ThemeContext'
import { OutlinedTextInput } from '../themed/OutlinedTextInput'
import { SceneHeader } from '../themed/SceneHeader'
import { SelectableRow } from '../themed/SelectableRow'

interface OwnProps {
  navigation: NavigationProp<'defaultFiatSetting'>
}
interface StateProps {
  supportedFiats: GuiFiatType[]
}
interface DispatchProps {
  onSelectFiat: (selectedDefaultFiat: string) => void
}
type Props = StateProps & DispatchProps & OwnProps & ThemeProps

interface State {
  supportedFiats: GuiFiatType[]
  selectedFiat: string
  searchTerm: string
}

export class DefaultFiatSettingComponent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      searchTerm: '',
      supportedFiats: props.supportedFiats,
      selectedFiat: ''
    }
  }

  handleSearchTermChange = (searchTerm: string): void => {
    this.setState({
      searchTerm
    })
  }

  renderFiatTypeResult = (data: FlatListItem<GuiFiatType>) => {
    const styles = getStyles(this.props.theme)
    const fiatCountry = FIAT_COUNTRY[data.item.value]
    if (!fiatCountry) {
      return null
    }

    return (
      <SelectableRow
        icon={fiatCountry.logoUrl ? <FastImage source={{ uri: fiatCountry.logoUrl }} style={styles.cryptoTypeLogo} /> : <View style={styles.cryptoTypeLogo} />}
        paddingRem={[0, 1]}
        // @ts-expect-error
        subTitle={s.strings[`currency_label_${data.item.value}`]}
        title={data.item.value}
        onPress={() => this.onSelectFiat(data.item)}
      />
    )
  }

  render() {
    const { theme } = this.props
    const styles = getStyles(this.props.theme)
    const filteredArray = this.props.supportedFiats.filter(entry => {
      return entry.label.toLowerCase().includes(this.state.searchTerm.toLowerCase())
    })

    return (
      <SceneWrapper avoidKeyboard background="theme" hasTabs={false}>
        {gap => (
          <View style={[styles.content, { marginBottom: -gap.bottom }]}>
            <SceneHeader withTopMargin title={s.strings.title_create_wallet_select_fiat} />
            <OutlinedTextInput
              autoCorrect={false}
              autoCapitalize="words"
              onChangeText={this.handleSearchTermChange}
              value={this.state.searchTerm}
              label={s.strings.fragment_wallets_addwallet_fiat_hint}
              returnKeyType="search"
              marginRem={[0, 1.75]}
              searchIcon
            />
            <FlashList
              style={styles.resultList}
              automaticallyAdjustContentInsets={false}
              contentContainerStyle={{ paddingBottom: gap.bottom }}
              data={filteredArray}
              estimatedItemSize={theme.rem(1.75)}
              keyboardShouldPersistTaps="handled"
              keyExtractor={this.keyExtractor}
              renderItem={this.renderFiatTypeResult}
            />
          </View>
        )}
      </SceneWrapper>
    )
  }

  onSelectFiat = ({ value: selectedFiat }: { value: string }) => {
    const { navigation } = this.props
    if (!this.isValidFiat(selectedFiat)) {
      Alert.alert(s.strings.fragment_create_wallet_select_valid)
    } else {
      this.setState({ selectedFiat })
      Keyboard.dismiss()
      this.props.onSelectFiat(selectedFiat)
      navigation.goBack()
    }
  }

  isValidFiat = (selectedFiat: string) => {
    const { supportedFiats } = this.state

    const isValid = supportedFiats.find(fiat => fiat.value === selectedFiat)

    return isValid
  }

  keyExtractor = (item: GuiFiatType) => `${item.label}${item.value}`
}

const getStyles = cacheStyles((theme: Theme) => ({
  content: {
    flex: 1,
    paddingTop: scale(5)
  },
  resultList: {
    flex: 1
  },
  cryptoTypeLogo: {
    width: theme.rem(2),
    height: theme.rem(2),
    borderRadius: theme.rem(1),
    marginLeft: theme.rem(0.25),
    backgroundColor: theme.backgroundGradientColors[1]
  }
}))

export const DefaultFiatSettingScene = connect<StateProps, DispatchProps, OwnProps>(
  state => ({
    supportedFiats: getSupportedFiats(getDefaultFiat(state))
  }),
  dispatch => ({
    onSelectFiat(selectedDefaultFiat) {
      dispatch(setDefaultFiatRequest(selectedDefaultFiat))
    }
  })
)(withTheme(DefaultFiatSettingComponent))
