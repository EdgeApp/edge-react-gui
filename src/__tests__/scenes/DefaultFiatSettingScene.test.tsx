import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { DefaultFiatSettingComponent } from '../../components/scenes/DefaultFiatSettingScene'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('DefaultFiatSettingComponent', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const actual = renderer.render(
      <DefaultFiatSettingComponent
        {...fakeSceneProps('defaultFiatSetting', {})}
        supportedFiats={[
          {
            label: 'Dollars',
            value: 'USD'
          }
        ]}
        onSelectFiat={async selectedDefaultFiat => {}}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
