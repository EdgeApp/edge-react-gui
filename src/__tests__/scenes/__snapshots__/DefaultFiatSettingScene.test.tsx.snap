// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultFiatSettingComponent should render with loading props 1`] = `
<SceneWrapper
  avoidKeyboard={true}
  background="theme"
  hasTabs={false}
>
  [Function]
</SceneWrapper>
`;
